consensus/babe/babe.go

```diff
@@ -221,16 +222,64 @@ func (b *Session) buildBlock(parent *types.BlockHeader, slot Slot) (*types.Block
 		return nil, err
 	}
 
+	// add block extrinsics
+	included, err := b.buildBlockExtrinsics(slot)
+	if err != nil {
+		return nil, err
+	}
+
 	// finalize block
+	log.Trace("build_block finalize block")
 	block, err := b.finalizeBlock()
 	if err != nil {
+		b.addToQueue(included)
 		return nil, err
 	}
 
 	block.Header.Number.Add(parent.Number, big.NewInt(1))
 	return block, nil
 }
 
+// buildBlockExtrinsics applies extrinsics to the block. it returns an array of included extrinsics.
+// for each extrinsic in queue, add it to the block, until the slot ends or the block is full.
+// if any extrinsic fails, it returns an empty array and an error.
+func (b *Session) buildBlockExtrinsics(slot Slot) ([]*tx.ValidTransaction, error) {
+	extrinsic := b.nextReadyExtrinsic()
+	included := []*tx.ValidTransaction{}
+
+	// TODO: check when block is full
+	for !hasSlotEnded(slot) && extrinsic != nil {
+		log.Trace("build_block", "applying extrinsic", extrinsic)
+		ret, err := b.applyExtrinsic(*extrinsic)
+		if err != nil {
+			return nil, err
+		}
+
+		// if ret == 0x0001, there is a dispatch error; if ret == 0x01, there is an apply error
+		if ret[0] == 1 || bytes.Equal(ret[:2], []byte{0, 1}) {
+			// TODO: specific error code checking
```

Comment:
Is there an issue open for this?

Reply:
https://github.com/ChainSafe/gossamer/issues/514